block/polka.go

```diff
@@ -82,58 +140,122 @@ func (builder PolkaBuilder) Insert(preVote SignedPreVote) {
+					// never used
 					preVotesForNil++
+
+					// This does let us skip nill blocks, but we
+					// already need to check for valid headers and
+					// signatures before any votes reach to this
+					// function. In other words, I think we should
+					// remove the whole if statement here.
 					continue
 				}
 				numPreVotes := preVotesForBlock[preVote.Block.Header]
 				numPreVotes++
 				preVotesForBlock[preVote.Block.Header] = numPreVotes
 			}
 
-			if preVotesForNil >= consensusThreshold {
-				highestPolkaFound = true
-				for _, preVote := range preVotes {
-					if preVote.Block == nil {
-						highestPolka.Block = preVote.Block
-						highestPolka.Round = preVote.Round
-						highestPolka.Height = preVote.Height
-						highestPolka.Signatories = append(highestPolka.Signatories, preVote.Signatory)
-						highestPolka.Signatures = append(highestPolka.Signatures, preVote.Signature)
-					}
-				}
-			}
+			polkaFound := false
 
+			// I needed to re-write this portion because it keeps the
+			// signatories and signatures from blocks at a lower
+			// height that also had enough prevotes to pass the
+			// consensusThreshold.
 			for blockHeader, numPreVotes := range preVotesForBlock {
 				if numPreVotes >= consensusThreshold {
-					highestPolkaFound = true
+					polkaFound = true
 					for _, preVote := range preVotes {
-						if preVote.Block != nil && preVote.Block.Header.Equal(blockHeader) {
+						if preVote.Block != nil &&
+							preVote.Block.Header.Equal(blockHeader) {
+
+							// if this happens you need to reset the
+							// signatories since you have just found a
+							// polka that has enough preVotes at a
+							// higher height.
+							if preVote.Height > highestPolka.Height {
+								highestPolka.Signatories = sig.Signatories{}
+								highestPolka.Signatures = sig.Signatures{}
+							}
+
 							highestPolka.Block = preVote.Block
 							highestPolka.Round = preVote.Round
 							highestPolka.Height = preVote.Height
-							highestPolka.Signatories = append(highestPolka.Signatories, preVote.Signatory)
-							highestPolka.Signatures = append(highestPolka.Signatures, preVote.Signature)
+							highestPolka.Signatories =
+								append(highestPolka.Signatories,
+									preVote.Signatory)
+							highestPolka.Signatures =
+								append(highestPolka.Signatures,
+									preVote.Signature)
 						}
 					}
 					break
 				}
 			}
+			if polkaFound {
+				continue
+			}
+
+			// if there are two different blocks voted for the same
+			// height then there are enough preVote to "think" there
+			// might be a polka, but the for loop above will realise
+			// none of the blocks have enough votes. So, we need to
+			// set the highestPolkaFound to false so we don't return a
+			// nil Polka and True. At the same time we don't want to
+			// set highestPolkaFound to false if we have already found
+			// a polka at a lower height.
+			if highestPolka.Block == nil {
+				highestPolkaFound = false
```

Comment:
This is incorrect. It’s valid to have a Polka for a nil Block. Please review the Tendermint consensus algorithm: PreVote and PreCommit for a nil Block is valid and, in fact, a correct part of what makes the algorithm safe and lively.